File: go/dijkstra/dijkstra.go
package dijkstra

func dijkstra(graph map[string]map[string]int, costs map[string]int, parents map[string]string) map[string]int {
	processed := make(map[string]struct{}, 0)
	node := findLowestCostNode(costs, processed)
	for node != "" {
		cost := costs[node]
		neighbors := graph[node]
		for n, c := range neighbors {
			newCost := cost + c
			if costs[n] > newCost {
				costs[n] = newCost
				parents[n] = node
			}
		}
		processed[node] = struct{}{}
		node = findLowestCostNode(costs, processed)
	}

	return costs
}

func findLowestCostNode(costs map[string]int, processed map[string]struct{}) string {
	lowestCost := 9999
	lowestCostNode := ""

	for node, cost := range costs {
		_, ok := processed[node]
		if cost < lowestCost && !ok {
			lowestCost = cost
			lowestCostNode = node
		}
	}
	return lowestCostNode
}

File: SwiftAlgorithms/SwiftAlgorithms/Dijkstra/Dijkstra.swift
//
//  Dijkstra.swift
//  SwiftAlgorithms
//
//  Created by ke on 2024/7/9.
//

import Foundation

class Dijkstra {
    static private var processed = Set<String>()
    
    static func dijkstra(_ graph: [String : [String : Int]], _ costs: inout [String : Int], _ parents: inout [String : String]) -> [String : Int] {
        var node = findLowestCostNode(costs)
        while let nodev = node {
            let cost = costs[nodev]!
            let neighbors = graph[nodev]!
            for (n, c) in neighbors {
                let newCost = cost + c
                if costs[n]! > newCost {
                    costs[n] = newCost
                    parents[n] = node
                 }
            }
            processed.insert(nodev)
            node = findLowestCostNode(costs)
        }
        
        return costs
    }
    
    static func findLowestCostNode(_ costs: [String : Int]) -> String? {
        var lowestCost = 9999
        var lowestCostNode: String?
        
        for (node, cost) in costs {
            if cost < lowestCost && !processed.contains(node){
                lowestCost = cost
                lowestCostNode = node
            }
        }
        
        return lowestCostNode
    }
}
